Add contributing guidelines Initial discussion (#3)
RISC-OS-Community/DME-Core: CONTRIBUTING.md

# Contributing to this project/repository
To contribute is fairly easy:

Until a proper RISC OS git tool is being released, we encourage you to use for your "local git" either a Windows computer or a Linux computer or Virtual Machine.

There are some very useful tips [here at the ROOL Forum](https://www.riscosopen.org/content/documents/git-cheatsheet)

## General guidelines
In general you'd want to:
- Git clone a repo from here to your local git

> - Create your branch using git branch (usually for branching naming you'd want to create the branch name as: <your nick>-<what type of changes you are applying>)

Reviewer: Is this right? Wouldn't you fork and the clone from there, then push to your fork and do a pull request? That's how we do it on the ROOL gitlab anyway.

Author: We can either fork and clone, but if we use branches is quicker to submit a pull request , just by experience. If we prefer forks then that's good for me. 

For the branching, just:

- git clone <repo>
- cd <repo>
- git branch mybranch-doing-this-change
- code / change things / run tests / git add+commit / git push -u origin mybranch-doing-this-change
- When you're finally done, just open a pull request either via git (CLI) or here on github (WEB) and tag the code-reviews-team 
- wait for the storm of comments, fix your code in your branch as required
- git add+commit / git push -u origin mybranch-doing-this-change
- resubmit your pull request (as I did for the code of conduct changes) so we have full history of those changes in a single place and with fixes
- done

But I am open to suggestions if we want to use a different approach :)

Author: For the "Do you need to do something with keys?"

I strongly recommend to use RSA keys (just add your public key if you haven't done it yet) that makes pushes quicker and also safer. Passwords are historically weak and, even if this is an Open Source community, you do want to avoid intruders to be able to impersonate you. 

If we all agree on using RSA keys I will add the procedure OR the links to the github procedures in the CONTRINUTING.md.

I am sorry I code for big companies so git is something I use on a daily base and like constantly, so I did not think of adding more details, my bad. Thanks for pointing it out btw!

Author: Ok added more details and background info so we can decide what is best.
Guys please let me know if we prefer Fork and branch in the fork or if we prefer to branch from the codebase. I am happy with both, thanks!
@skymandr @InteractiveYelp ^^^

> - Git clone a repo from here to your local git
> - Create your branch using git branch (usually for branching naming you'd want to create the branch name as: <your nick>-<what type of changes you are applying>)
> - When done you have to submit your branch for review requesting a code review to the Code Reviews Team
> - If you can please make sure your commit messages are clear and meaningful to help reviewers to understand what you were trying to achieve
> - When possible try to make small commits so that the review process can be quicker
> - Make sure you keep the documentation up-to-date with your changes
> - Where applicable stick to RISC OS Style guide
> - Make sure you've tested your changes before submitting a Pull Request

Reviewer 2: ```suggestion
- Git clone a repo from here to your local git
- Create your branch using git branch (we suggest that you name your branches following this pattern: <your nick>-<what type of changes you are applying>)
- Make sure your commit messages are clear and meaningful to help reviewers to understand what you were trying to achieve
- When possible try to make small commits and Pull Requests so that the review process can be quicker
- Make sure you keep the documentation up-to-date with your changes
- Where applicable stick to RISC OS Style guide
- Make sure you've tested your changes before submitting a Pull Request
- Submit your branch for review when it is done, do this by making a Pull Request and requesting a review from the Code Reviews Team
```

3) Reviewing/Improving requirements and documentation
4) Helping us with the graphics/sound that certain projects and repositories may require
5) Helping us testing the code on as many RISC OS platforms and OS versions as possible (within the supported range)
6) Providing us constructive feedback on the work/projects/repositories/requirements/designs

Next chapters will help who is completely new to this process, introduce a common dictionary of technical terms to make sure we are all on the same page and finally provide instructions on how to submit back your contribution.

Thanks for considering helping us and we hope you'll have fun!

## Is this your first contribution ever?

> If this is your first ever contribution to an Open Source project then don't be afraid, here there is a free course for you to follow that will give you a lot of good information: [How to contribute to an Open Source project on GitHub](https://app.egghead.io/playlists/how-to-contribute-to-an-open-source-project-on-github)

Reviewer 2: ```suggestion
If this is your first ever contribution to an Open Source project then don't be afraid. Here is a free course for you to follow that will give you a lot of useful information: [How to contribute to an Open Source project on GitHub](https://app.egghead.io/playlists/how-to-contribute-to-an-open-source-project-on-github)
```

## Submitting code

> To submit your changes we use Pull Requests (if you used GitLab instead they are called there Merge Requests). Any code change should be submitted as a pull request, no other ways are allowed, sorry. Read below for the details and the guidelines.

Reviewer 2: ```suggestion
To submit our changes we use Pull Requests (if you are used to GitLab you may know them as Merge Requests). Any code change should be submitted as a pull request, no other ways are allowed, sorry. Read below for the details and guidelines.
```

  - we suggest that you name your branches following this pattern: `your_nick-what_type_of_changes_you_are_applying` with no spaces, you can use either dash or underscore to your preference
* Start adding your changes
  - Where applicable stick to RISC OS Style guide

>     - If you don't have the guide and would lik eto get your copy you can fin dinfo/help [here](https://www.riscosopen.org/wiki/documentation/show/Style%20Guide)

Reviewer 2: ```suggestion
    - If you don't have the guide and would like to get your copy you can find info/help [here](https://www.riscosopen.org/wiki/documentation/show/Style%20Guide)
```

Reviewer 2: I will have another full read-through tonight, but it looks good I think.

[Tools to help using git with BBC BASIC](https://gitlab.riscosopen.org/Support/RepoTools)

# Extra info for everyone

> If you fear having to learn git or if you just want to help us with reviewing documentation etc you can also use github via the github application for iOS and Android available for both the Apple iPad and iPhone and Android Tablets and Phones.
> The Android/Apple github application makes it very easy to interact with us and provide your contribution without the need to learn git!
> Have a look on your device on-line store for the GitHub application published by GitHub.
> 
> So no excuses ;-)

Reviewer 2: ```suggestion
If you fear having to learn git or if you just want to help us with reviewing documentation etc. you can also use GitHub via the GitHub application, available for both the Apple iPad and iPhone and Android Tablets and Phones.
The Android/Apple GitHub application makes it very easy to interact with us and provide your contribution without the need to learn git!
Have a look in your device' on-line store for the GitHub application published by GitHub.

So no excuses! ;-)
```
